coverage.c

```diff
@@ -0,0 +1,590 @@
+                print_error_errno("coverage", "can't parse region \"%s\"", opt_reg);
+                status = EXIT_FAILURE;
+                goto coverage_end;
+            }
+        }
+    }
+
+    if (opt_print_tabular && opt_print_header)
+        fputs("#CHROM\tSTART\tEND\tN_READS\tN_COVERED_BASES\tPERCENT_COVERED\tAVG_COV\tAVG_BASEQ\tAVG_MAPQ\n", file_out);
+
+
+    h = data[0]->hdr; // easy access to the header of the 1st BAM
+    covered_tids = calloc(h->n_targets, sizeof(bool));
+    stats_aux_t *stats = calloc(1, sizeof(stats_aux_t));
+
+    bool reg_sets_end = false;
+    if (opt_reg) {
+        stats->beg = data[0]->iter->beg; // and to the parsed region coordinates
+        stats->end = data[0]->iter->end;
+        stats->tid = data[0]->iter->tid;
+        stats->bin_width = (stats->end-stats->beg) / opt_n_bins;
+        reg_sets_end = stats->end != INT_MAX;
+    } else {
+        stats->tid = -1;
+    }
+
+    int current_bin;
+
+    // the core multi-pileup loop
+    mplp = bam_mplp_init(n_bam_files, read_bam, (void**)data); // initialization
+    if (max_depth > 0)
+        bam_mplp_set_maxcnt(mplp,max_depth);  // set maximum coverage depth
+    else if (!max_depth)
+        bam_mplp_set_maxcnt(mplp,INT_MAX);
+
+
+    // Extra info for histogram and coverage counting
+    uint32_t *hist = (uint32_t*) calloc(opt_n_bins, sizeof(uint32_t));
+    //uint32_t **hists_for_html = (uint32_t**) calloc(opt_n_bins*h->n_targets, sizeof(uint32_t));
+    //stats_aux_t **stats_for_html = (stats_aux_t*) calloc(h->n_targets, sizeof(stats_aux_t));
+    
+    n_plp = (int*) calloc(n_bam_files, sizeof(int*)); // n_plp[i] is the number of covering reads from the i-th BAM
+    plp = (const bam_pileup1_t**) calloc(n_bam_files, sizeof(bam_pileup1_t*)); // plp[i] points to the array of covering reads (internal in mplp)
+    while ((ret=bam_mplp_auto(mplp, &tid, &pos, n_plp, plp)) > 0) { // come to the next covered position
+        if (tid != stats->tid) {
+            if (stats->tid >= 0) {
+                set_read_counts(data, stats, n_bam_files);
+                if (opt_print_histogram) {
+                    print_hist(file_out, h, stats, hist, opt_n_bins);
+                    fputc('\n', file_out);
+                } else if (opt_print_tabular) {
+                    print_tabular_line(file_out, h, stats);
+                }
+
+                for (i = 0; i < n_bam_files && data[i]; ++i) {
+                    data[i]->n_reads = 0;
+                    data[i]->n_selected_reads = 0;
+                }
+
+                memset(stats, 0, sizeof(stats_aux_t));
+
+                if (compute_histogram)
+                    memset(hist, 0, opt_n_bins*sizeof(uint32_t));
+            }
+
+            stats->tid = tid;
+            covered_tids[tid] = true;
+            if (!reg_sets_end)
+                stats->end = h->target_len[tid];
+
+            if (compute_histogram) 
+                stats->bin_width = (stats->end-stats->beg) / opt_n_bins;
```

Comment:
This needs a check to ensure bin_width never hits zero.  Maybe adjust opt_n_bins if it does to guarantee it cannot happen.

Currently with a tiny region the current_bin calculation below gets a divide by zero and crashes.

Reply:
Fixed by ensuring that a bin is at minimum one base